types/uuid.go

```diff
@@ -23,57 +23,58 @@ import (
 	"io"
 )
 
-type UUID struct {
-	mostSigBits  uint64
-	leastSigBits uint64
-}
+type UUID [16]byte
```

Comment:
> What's the positive impact of refactored UUID?

Prior to this change there were litter allocations due to `UUID.String()` that happened on each operation as a part of map lookups. With this change, UUIDs can be directly used as map keys avoiding the litter.

> I didn't run any benchmarks, but my gut feeling is that comparing 2 64bit integers is more efficient than comparing a byte array of 16.

If you mean map key comparison, you should compare "struct-to-string conversion -> string hash calculation -> string comparison" with "byte array hash calculation -> byte array comparison". I'd be surprised with the former being more efficient.

> I believe the former is also more space efficient.

How two 8 bytes field struct can be more space efficient than 16 bytes array?

Reply:
As far as I've checked comparing 2 64bit  integers is 4x faster than comparing 16 bytes. My point is, the important change in this PR is using UUID as is without converting it to a string first. That can be done using the previous implementation too.

Reply:
> As far as I've checked comparing 2 64bit integers is 4x faster than comparing 16 bytes.

Sounds interesting. Could you share more details? From CPU standpoint, comparing 16 bytes should be fairly efficient (see https://github.com/hazelcast/hazelcast-go-client/pull/564#discussion_r642257227) and I'm curious about the difference with the two `uint64` fields (struct) approach in terms of the emitted instructions.

> My point is, the important change in this PR is using UUID as is without converting it to a string first. That can be done using the previous implementation too.

It's certainly possible to use structs with comparable fields as map keys. I was following the same approach as https://github.com/google/uuid library here. If there is a reason to change to structs, I can do that.

Reply:
I did a quick experiment and checked the compilation result.

The following Go code checks two structs for equality:
```go
//go:noinline
type UUID0 struct {
	mostSigBits  uint64
	leastSigBits uint64
}

func newUUID0() UUID0 {
	return UUID0{mrand.Uint64(), mrand.Uint64()}
}

func main() {
	uuid1 := newUUID0()
	uuid2 := newUUID0()
	if uuid1 == uuid2 {
		fmt.Println("Ka-boom!")
	}
}
```
and here are the emitted instructions:
```
TEXT main.main(SB) /home/puzpuzpuz/projects/go-experiments/main.go
  main.go:18		0x498080		64488b0c25f8ffffff	MOVQ FS:0xfffffff8, CX			
  main.go:18		0x498089		483b6110		CMPQ 0x10(CX), SP			
  main.go:18		0x49808d		0f86ad000000		JBE 0x498140				
  main.go:18		0x498093		4883ec68		SUBQ $0x68, SP				
  main.go:18		0x498097		48896c2460		MOVQ BP, 0x60(SP)			
  main.go:18		0x49809c		488d6c2460		LEAQ 0x60(SP), BP			
  main.go:19		0x4980a1		e85affffff		CALL main.newUUID0(SB)			
  main.go:21		0x4980a6		488b0424		MOVQ 0(SP), AX				
  main.go:21		0x4980aa		4889442448		MOVQ AX, 0x48(SP)			
  main.go:21		0x4980af		488b4c2408		MOVQ 0x8(SP), CX			
  main.go:21		0x4980b4		48894c2440		MOVQ CX, 0x40(SP)			
  main.go:20		0x4980b9		e842ffffff		CALL main.newUUID0(SB)			
  main.go:21		0x4980be		488b442408		MOVQ 0x8(SP), AX			
  main.go:21		0x4980c3		488b4c2448		MOVQ 0x48(SP), CX			
  main.go:21		0x4980c8		48390c24		CMPQ CX, 0(SP)				
  main.go:21		0x4980cc		750a			JNE 0x4980d8				
  main.go:21		0x4980ce		488b4c2440		MOVQ 0x40(SP), CX			
  main.go:21		0x4980d3		4839c1			CMPQ AX, CX				
  main.go:21		0x4980d6		740a			JE 0x4980e2				
  main.go:21		0x4980d8		488b6c2460		MOVQ 0x60(SP), BP			
  main.go:21		0x4980dd		4883c468		ADDQ $0x68, SP				
  main.go:21		0x4980e1		c3			RET			
...		
```

Reply:
Less/More instructions don't always correspond to better efficiency. You could have a few instructions which loops millions of times which would run slower than a few hundred lines of code. In the code below, `U1` corresponds to the previous implementation and `U2` corresponds to implementation in this PR:
```
package main

import (
	"fmt"
	"time"
)

type U1 struct {
	low uint64
	hi uint64
}

type U2 [16]byte

func do1(times int) time.Duration {
	u1 := U1{}
	u2 := U1{}
	c := 0
	tic := time.Now()
	for i := 0; i < times; i++ {
		if u1 == u2 {
			c++
		}
	}
	toc := time.Now()
	fmt.Println(c)
	return toc.Sub(tic)
}

func do2(times int) time.Duration {
	u1 := U2{}
	u2 := U2{}
	c := 0
	tic := time.Now()
	for i := 0; i < times; i++ {
		if u1 == u2 {
			c++
		}
	}
	toc := time.Now()
	fmt.Println(c)
	return toc.Sub(tic)
}

func main() {
	took := do1(100_000_000)
	fmt.Println("d1 Took:", took.Milliseconds())
	took = do2(100_000_000)
	fmt.Println("d2 Took:", took.Milliseconds())
}
```
`d1` is consistently 4x faster than `d2` for me.

Reply:
> Less/More instructions don't always correspond to better efficiency.

Sure, there are a lot of other things that matter: memory access patterns, ILP-friendliness (data dependency), and so on. But in this particular case, I don't see a lot of difference in the instructions.

Thanks for sharing your benchmark. I'll check your code tomorrow.

Reply:
I've checked your code sample and, as always, microbenchmarks are really tricky. The thing is that the Go compiler is smart enough to understand that your structs are equal and removes the `if u1 == u2` check in the `do1` function completely. Here is the compilation result:
```
  main.go:20		0x497dc2		488b9c2498000000	MOVQ 0x98(SP), BX			
  main.go:20		0x497dca		31f6			XORL SI, SI				
  main.go:20		0x497dcc		eb03			JMP 0x497dd1				
  main.go:22		0x497dce		48ffc6			INCQ SI					
  main.go:20		0x497dd1		4839f3			CMPQ SI, BX				
  main.go:20		0x497dd4		7ff8			JG 0x497dce				
  main.go:20		0x497dd6		4889742440		MOVQ SI, 0x40(SP)			
```

L20 here stands for the `for` loop, L22 is the `c++` increment operator.

In case of arrays, the compiler is not smart enough, so it does the check for each loop iteration. So, you were comparing a raw `c++` loop for structs with the check-included loop for byte arrays.

To make the program closer to what we want to benchmark, we need to use dynamic values for both struct fields and byte arrays. Here is the modified program:
```go
package main

import (
	"fmt"
	"math/rand"
	"time"
)

type U1 struct {
	low uint64
	hi  uint64
}

type U2 [16]byte

func newU1() U1 {
	return U1{rand.Uint64(), rand.Uint64()}
}

func newU2() U2 {
	u := U2{}
	for i := 0; i < 16; i++ {
		u[i] = byte(rand.Int())
	}
	return u
}

func do1(times int) (time.Duration, U1, U1) {
	u1 := newU1()
	u2 := newU1()
	c := 0
	tic := time.Now()
	for i := 0; i < times; i++ {
		if u1 == u2 {
			c++
		}
	}
	toc := time.Now()
	fmt.Println(c)
	return toc.Sub(tic), u1, u2
}

func do2(times int) (time.Duration, U2, U2) {
	u1 := newU2()
	u2 := newU2()
	c := 0
	tic := time.Now()
	for i := 0; i < times; i++ {
		if u1 == u2 {
			c++
		}
	}
	toc := time.Now()
	fmt.Println(c)
	return toc.Sub(tic), u1, u2
}

func main() {
	rand.Seed(time.Now().UnixNano())
	took, u1_1, u1_2 := do1(100_000_000)
	fmt.Println("d1 Took:", took.Milliseconds())
	fmt.Printf("d1 result: %v, %v\n", u1_1, u1_2)
	took, u2_1, u2_2 := do2(100_000_000)
	fmt.Println("d2 Took:", took.Milliseconds())
	fmt.Printf("d2 result: %v, %v\n", u2_1, u2_2)
}
```

This one gives the something like the following result on my machine:
```
$ go run main.go 
0
d1 Took: 52
d1 result: {2868247082766820141 16168540106036637080}, {5421879556549508384 6620505235924796315}
0
d2 Took: 78
d2 result: [108 100 83 112 149 251 175 47 75 125 188 14 82 111 146 33], [72 72 82 94 87 220 126 226 217 154 202 133 62 146 231 145]
```

So, the real difference between struct vs. small byte array equality is around 50%, not 300% (let's ignore potential HW effects involved in the microbenchmark for the sake of simplicity). If you're interested in seeing the compilation results for both loops, please let me know and I'll post them here.

I don't think such difference makes any weather (the same would apply even the difference would be 4x). But in order to get some progress with the code review, I'm going to rewrite the `UUID` to be a struct, not a byte array.

Reply:
Did the rewrite in b186bdecd08e2cca1e84b7c4728ebda2b140aea1

Reply:
My point was not even which one was faster, but the main improvement in this PR is that the UUID itself is used as the comparison key. Thanks for the update.